ipa-core/src/protocol/ipa_prf/validation_protocol/validation.rs

```diff
@@ -213,6 +219,55 @@ impl BatchToVerify {
             challenges_for_prover_right.collect(),
         )
     }
+
+    /// This function computes and outputs the final `p_r_from_left_prover * q_r_from_left_prover` value.
+    async fn p_and_q_r_check<C, U, V>(
+        &self,
+        ctx: &C,
+        challenges_for_left_prover: &[Fp61BitPrime],
+        challenges_for_right_prover: &[Fp61BitPrime],
+        u_from_right_prover: U,
+        v_from_left_prover: V,
+    ) -> Result<Fp61BitPrime, Error>
+    where
+        C: Context,
+        U: Iterator<Item = Fp61BitPrime> + Send,
+        V: Iterator<Item = Fp61BitPrime> + Send,
+    {
+        const LRF: usize = LargeProofGenerator::RECURSION_FACTOR;
+        const SRF: usize = SmallProofGenerator::RECURSION_FACTOR;
+
+        // compute p_r
+        let p_r_right_prover = recursively_compute_final_check::<_, _, LRF, SRF>(
+            stream::iter(u_from_right_prover),
+            challenges_for_right_prover,
+            self.p_mask_from_right_prover,
+        )
+        .await;
+        // compute q_r
+        let q_r_left_prover = recursively_compute_final_check::<_, _, LRF, SRF>(
+            stream::iter(v_from_left_prover),
+            challenges_for_left_prover,
+            self.q_mask_from_left_prover,
+        )
+        .await;
+
+        // send to the left
+        let communication_ctx = ctx.set_total_records(1);
+
+        let send_channel =
+            communication_ctx.send_channel::<Fp61BitPrime>(ctx.role().peer(Direction::Right));
+        let receive_channel =
+            communication_ctx.recv_channel::<Fp61BitPrime>(ctx.role().peer(Direction::Left));
+
+        let ((), q_r_right_prover) = try_join(
+            send_channel.send(RecordId::FIRST, q_r_left_prover),
```

Comment:
Since party P_{i-1} is the one with q(r), it needs to reveal it to P_{i+1}. That would be *left*, not *right*. That would be sending it back to the prover again.